File: app/OculusKit/Sources/OculusUI/PreviewDOM.swift
import Foundation
#if canImport(WebKit)
import WebKit
#endif

/// Performing an agent's DOM operations in the preview the user already has open.
///
/// The daemon has no browser. A dev server's HTML is inert, and for any client-rendered app the
/// markup it serves is an empty shell — everything real appears only after the page's own JavaScript
/// runs. The two ways to give the daemon a DOM both mean shipping Chromium to every user.
///
/// The app already has a browser engine, so the work happens here: the daemon broadcasts an ask,
/// whichever client has that session's preview open performs it, and the answer goes back.
///
/// The limitation is inherent and is stated to the agent rather than hidden: nothing works while
/// nobody is looking. A refusal an agent can act on beats a confident answer about a page that was
/// never rendered.

/// Tracks which web view is showing which session's preview.
///
/// Weak references throughout: a registry that keeps web views alive would leak one per Design sheet
/// opened, and — worse — a retained but detached view would answer an agent about a page nobody is
/// looking at any more.
public final class PreviewDOMRegistry {
    public static let shared = PreviewDOMRegistry()

    #if canImport(WebKit)
    private final class WeakView {
        weak var view: WKWebView?
        init(_ v: WKWebView) { view = v }
    }
    private var views: [String: WeakView] = [:]
    #endif
    private let lock = NSLock()

    #if canImport(WebKit)
    public func register(sessionID: String, view: WKWebView) {
        guard !sessionID.isEmpty else { return }
        lock.lock(); views[sessionID] = WeakView(view); lock.unlock()
    }

    public func unregister(sessionID: String) {
        guard !sessionID.isEmpty else { return }
        lock.lock(); views.removeValue(forKey: sessionID); lock.unlock()
    }

    /// The live web view for a session, if one is still on screen.
    public func view(for sessionID: String) -> WKWebView? {
        lock.lock(); defer { lock.unlock() }
        guard let box = views[sessionID] else { return nil }
        if box.view == nil {
            views.removeValue(forKey: sessionID) // the sheet closed; drop the empty box
            return nil
        }
        return box.view
    }
    #endif

    #if canImport(WebKit)
    /// Web views this app created for an agent, rather than for a person to look at.
    ///
    /// Held STRONGLY, unlike the ones above: nothing else is keeping them alive. Evicted after an
    /// idle period so a day of agent work does not accumulate one renderer per session.
    private var headless: [String: (view: WKWebView, lastUsed: Date)] = [:]

    /// How long an agent-created view survives without use.
    private static let headlessIdle: TimeInterval = 10 * 60

    /// A view for this session, preferring one a PERSON already has open.
    ///
    /// The visible one is preferred because it is the page the user is actually looking at — the
    /// agent and the human should be seeing the same thing where possible.
    public func anyView(for sessionID: String) -> WKWebView? {
        if let visible = view(for: sessionID) { return visible }
        lock.lock(); defer { lock.unlock() }
        evictIdleLocked()
        guard let entry = headless[sessionID] else { return nil }
        headless[sessionID] = (entry.view, Date())
        return entry.view
    }

    /// Registers a view this app created for agent use.
    public func adoptHeadless(sessionID: String, view: WKWebView) {
        guard !sessionID.isEmpty else { return }
        lock.lock(); defer { lock.unlock() }
        evictIdleLocked()
        headless[sessionID] = (view, Date())
    }

    /// Drops an agent-created view — when its session ends, or its page failed to load.
    public func dropHeadless(sessionID: String) {
        lock.lock(); defer { lock.unlock() }
        headless.removeValue(forKey: sessionID)
    }

    private func evictIdleLocked() {
        let cutoff = Date().addingTimeInterval(-Self.headlessIdle)
        for (id, entry) in headless where entry.lastUsed < cutoff {
            headless.removeValue(forKey: id)
        }
    }
    #endif

    /// Whether any web view is currently showing this session's preview.
    public func isShowing(sessionID: String) -> Bool {
        #if canImport(WebKit)
        return view(for: sessionID) != nil
        #else
        return false
        #endif
    }
}

#if canImport(WebKit)
/// Builds an offscreen web view for a session's preview, so an agent can look at the page without a
/// person having the Design sheet open.
///
/// The alternative was to have the daemon ask the app to OPEN that sheet, which works and is wrong:
/// an agent would be yanking a window in front of whatever its owner was doing, and the moment that
/// is possible it is also a thing a misbehaving agent can do repeatedly. This renders the same page
/// with the same scripts and interrupts nobody.
///
/// The frame is explicit and non-zero on purpose. A WKWebView with a zero frame still runs
/// JavaScript, so the temptation is to leave it unsized — but it does not LAY OUT, and the snapshot
/// walks the page asking each element for its client rects. Unsized, every element reports no boxes,
/// the visibility check drops all of them, and the agent is told a fully working page is empty. That
/// is the worst possible answer: confidently wrong rather than unavailable.
@MainActor
public enum HeadlessPreview {
    /// A conventional desktop viewport. Responsive layouts branch on width, so a page rendered at
    /// 320pt would answer questions about the mobile layout while the person asking is looking at
    /// the desktop one.
    static let viewport = CGSize(width: 1280, height: 900)

    /// How long to wait for the page before giving up and saying so.
    static let loadTimeout: TimeInterval = 20

    /// Creates a view, loads `url`, and returns it once the page has finished loading.
    public static func make(url: URL, tunnel: PreviewSchemeHandler?) async -> WKWebView? {
        let cfg = WKWebViewConfiguration()
        if let tunnel {
            cfg.setURLSchemeHandler(tunnel, forURLScheme: previewTunnelScheme)
        }
        let wv = WKWebView(frame: CGRect(origin: .zero, size: viewport), configuration: cfg)
        let waiter = LoadWaiter()
        wv.navigationDelegate = waiter
        wv.load(URLRequest(url: url))
        let loaded = await waiter.wait(timeout: loadTimeout)
        // The delegate is cleared either way: it exists only to observe the first load, and leaving
        // it attached would keep a waiter alive for the life of the view.
        wv.navigationDelegate = nil
        return loaded ? wv : nil
    }
}

/// Resumes once a web view finishes (or fails) its first navigation.
private final class LoadWaiter: NSObject, WKNavigationDelegate {
    private var continuation: CheckedContinuation<Bool, Never>?
    private var settled = false

    func wait(timeout: TimeInterval) async -> Bool {
        await withCheckedContinuation { (c: CheckedContinuation<Bool, Never>) in
            continuation = c
            // A dev server that accepts the connection and then never answers would otherwise park
            // this forever, and the agent with it.
            Task { @MainActor in
                try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                self.settle(false)
            }
        }
    }

    private func settle(_ ok: Bool) {
        guard !settled else { return } // a failure after a success, or the timeout after either
        settled = true
        continuation?.resume(returning: ok)
        continuation = nil
    }

    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) { settle(true) }
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) { settle(false) }
    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) { settle(false) }
}
#endif

/// Reads the live DOM into an outline an agent can reason about, stamping each listed element with a
/// ref.
///
/// Refs rather than CSS selectors, in both directions. A selector the agent composes could reach any
/// element on the page, including ones it was never shown; a ref can only name something a snapshot
/// actually returned. It also survives a class name changing underneath, which is exactly what
/// happens while an agent is editing the styles.
public let previewSnapshotJS = """
(function() {
  var MAX = 200;
  // Every snapshot opens a new EPOCH, and the epoch is part of each ref.
  //
  // Without it, refs restart at e1 on every snapshot, so a ref held from an earlier snapshot can
  // resolve to a completely different element after the page re-renders — the agent clicks "Save"
  // and hits "Delete", with nothing anywhere reporting a problem. A silent mis-click is the most
  // expensive failure this feature can have, and it is only detectable if the ref itself carries
  // which snapshot it came from.
  var epoch = (window.__irEpoch || 0) + 1;
  window.__irEpoch = epoch;
  var sel = 'a,button,input,select,textarea,summary,label,[role],[onclick],h1,h2,h3,h4';
  var seen = 0, out = [];
  function visible(el) {
    if (!el.getClientRects().length) return false;
    var s = getComputedStyle(el);
    return s.visibility !== 'hidden' && s.display !== 'none' && s.opacity !== '0';
  }
  // Fields whose CONTENTS must never be reported, whatever else is true of them.
  //
  // Checked in one place because there are two ways a value escapes and only one is obvious. The
  // explicit `value` branch below is easy to remember; the accessible NAME is not — it falls back
  // through label, placeholder and title to el.value, so an unlabelled password field reported its
  // own contents as its name. A real browser found that; reading the source did not.
  function sensitive(el) {
    var type = (el.getAttribute('type') || '').toLowerCase();
    if (type === 'password') return true;
    var hint = ((el.getAttribute('name') || '') + ' ' + (el.id || '') + ' ' +
                (el.getAttribute('autocomplete') || '')).toLowerCase();
    return /pass|secret|token|otp|cvv|cvc|card|cc-|ssn/.test(hint);
  }
  function name(el) {
    var t = (el.getAttribute('aria-label') || el.getAttribute('placeholder') ||
             el.getAttribute('title') || el.innerText ||
             (sensitive(el) ? '' : el.value) || '').trim();
    return t.length > 120 ? t.slice(0, 120) + '…' : t;
  }
  document.querySelectorAll(sel).forEach(function(el) {
    if (seen >= MAX || !visible(el)) return;
    seen++;
    var ref = 's' + epoch + 'e' + seen;
    el.setAttribute('data-ir-ref', ref);
    var e = { ref: ref, tag: el.tagName.toLowerCase() };
    var role = el.getAttribute('role'); if (role) e.role = role;
    var n = name(el); if (n) e.name = n;
    if (el.tagName === 'INPUT' || el.tagName === 'TEXTAREA' || el.tagName === 'SELECT') {
      e.type = el.getAttribute('type') || el.tagName.toLowerCase();
      // A snapshot goes into the agent's context AND into the durable transcript, where a secret
      // captured once is retained long after the session ends.
      if (sensitive(el)) {
        if (el.value) e.value = '[withheld]'; // say a value EXISTS without saying what it is
      } else if (el.value) {
        e.value = String(el.value).slice(0, 120);
      }
    }
    if (el.disabled) e.disabled = true;
    out.push(e);
  });
  return JSON.stringify({
    url: location.href,
    title: document.title,
    truncated: seen >= MAX,
    elements: out
  });
})()
"""

/// Finds a snapshotted element by ref, without ever putting the ref into a selector.
///
/// Building `'[data-ir-ref="' + ref + '"]'` would hand agent-supplied text to the CSS parser, and a
/// ref carrying a quote either breaks the selector or reaches further than intended. Comparing the
/// attribute in a loop removes the parser from the path entirely — bounded work, since a snapshot
/// stamps at most a couple of hundred elements.
private let previewFindByRefJS = """
      var m = /^s(\\d+)e\\d+$/.exec(ref);
      if (!m) return JSON.stringify({ error: 'Malformed ref ' + ref + '. Use a ref from preview_snapshot.' });
      if (Number(m[1]) !== (window.__irEpoch || 0)) {
        return JSON.stringify({ error: 'Stale ref ' + ref + ' — the page has been snapshotted again since. Take a fresh preview_snapshot and use its refs.' });
      }
      var el = null, all = document.querySelectorAll('[data-ir-ref]');
      for (var i = 0; i < all.length; i++) {
        if (all[i].getAttribute('data-ir-ref') === ref) { el = all[i]; break; }
      }
      if (el && !el.isConnected) el = null; // detached by a re-render since the snapshot
"""

/// Clicks a previously-snapshotted element.
///
/// The ref is bound to a JS variable ONCE, encoded, and every later use concatenates that variable.
/// Interpolating it a second time — even into an error message — is how it ends up outside a string
/// literal: `error: 'No element \\(ref).'` looks like a message and is an injection point.
public func previewClickJS(ref: String) -> String {
    """
    (function() {
      var ref = \(jsString(ref));
    \(previewFindByRefJS)
      if (!el) return JSON.stringify({ error: 'No element ' + ref + '. The page may have re-rendered — take a fresh preview_snapshot.' });
      el.scrollIntoView({ block: 'center' });
      el.click();
      return JSON.stringify({ clicked: ref, url: location.href });
    })()
    """
}

/// Types into a previously-snapshotted input.
public func previewFillJS(ref: String, value: String) -> String {
    """
    (function() {
      var ref = \(jsString(ref)), value = \(jsString(value));
    \(previewFindByRefJS)
      if (!el) return JSON.stringify({ error: 'No element ' + ref + '. The page may have re-rendered — take a fresh preview_snapshot.' });
      if (!('value' in el)) return JSON.stringify({ error: 'Element ' + ref + ' is a ' + el.tagName.toLowerCase() + ', which cannot be typed into.' });
      el.focus();
      // React (and anything else tracking the value internally) ignores a plain assignment: it
      // installs its own value setter, so the change has to go through the NATIVE one and then be
      // announced with the events a framework listens for. Without this the field looks filled on
      // screen and the app's state never changes.
      var proto = el instanceof HTMLTextAreaElement ? HTMLTextAreaElement.prototype : HTMLInputElement.prototype;
      var setter = Object.getOwnPropertyDescriptor(proto, 'value');
      if (setter && setter.set) { setter.set.call(el, value); } else { el.value = value; }
      el.dispatchEvent(new Event('input', { bubbles: true }));
      el.dispatchEvent(new Event('change', { bubbles: true }));
      return JSON.stringify({ filled: ref });
    })()
    """
}

/// Encodes a Swift string as a JavaScript literal.
///
/// Every value an agent supplies is interpolated into a script, so this is the boundary that keeps a
/// ref or a form value from becoming code. JSONSerialization rather than hand-rolled escaping,
/// because JSON string syntax is a subset of JavaScript's and it already handles quotes, backslashes,
/// newlines and the unicode line separators that break naive escapers.
public func jsString(_ s: String) -> String {
    if let data = try? JSONSerialization.data(withJSONObject: [s], options: []),
       let arr = String(data: data, encoding: .utf8) {
        return String(arr.dropFirst().dropLast()) // ["x"] -> "x"
    }
    return "\"\"" // unencodable input becomes an empty string, never raw text
}

File: daemon/hub/preview_dom.go
package hub

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/howlerops/oculus/daemon/mcp"
	"github.com/howlerops/oculus/daemon/protocol"
)

// Letting an agent look at, and act on, the page it is building.
//
// The daemon has no DOM. It speaks HTTP, and a dev server's HTML is inert — for any client-rendered
// app the fetched markup is `<div id="root"></div>` and everything real happens after the page's own
// JavaScript runs. So an agent cannot verify its own UI work from here, and the two obvious ways to
// give the daemon a DOM both cost a Chromium dependency for every user whether they use it or not.
//
// The app already has a browser engine. So the daemon asks it: the request goes out to connected
// clients, whichever one has that session's preview open performs the operation in its own web view,
// and the answer comes back. Same shape as an approval — broadcast, wait on a keyed waiter, time out.
//
// The limitation is narrower than it first was. The app renders the page whether or not a person has
// the preview open: if nobody does, it builds an offscreen web view for the purpose, shows nothing
// and takes no focus. What remains is that SOME app has to be connected — there is no renderer
// otherwise — and that case still refuses clearly rather than guessing, because a wrong answer about
// what a page displays is worse than no answer.
//
// Deliberately not "the daemon asks the app to open the Design sheet". That works, and it means an
// agent can throw a window in front of whatever its owner was doing — which is a capability worth
// not having.

// previewDOMTimeout bounds one DOM round trip. Generous enough for a slow page on a phone over a
// relay, short enough that an agent is not parked on a client that has gone away.
const previewDOMTimeout = 20 * time.Second

// previewDOMWaiters correlates an outstanding ask with the client's eventual answer.
type previewDOMWaiters struct {
	mu sync.Mutex
	ch map[string]chan protocol.PreviewDOMResult
}

func newPreviewDOMWaiters() *previewDOMWaiters {
	return &previewDOMWaiters{ch: map[string]chan protocol.PreviewDOMResult{}}
}

func (w *previewDOMWaiters) open(id string) chan protocol.PreviewDOMResult {
	// Buffered so a client answering after the deadline never blocks its own read loop on a waiter
	// nobody is reading any more.
	c := make(chan protocol.PreviewDOMResult, 1)
	w.mu.Lock()
	w.ch[id] = c
	w.mu.Unlock()
	return c
}

func (w *previewDOMWaiters) close(id string) {
	w.mu.Lock()
	delete(w.ch, id)
	w.mu.Unlock()
}

// resolve delivers an answer. The FIRST answer wins: several devices may have the same preview open,
// and the agent asked one question.
func (w *previewDOMWaiters) resolve(res protocol.PreviewDOMResult) {
	w.mu.Lock()
	c, ok := w.ch[res.RequestID]
	if ok {
		delete(w.ch, res.RequestID)
	}
	w.mu.Unlock()
	if ok {
		c <- res
	}
}

// askPreviewDOM performs one operation in whichever connected app has this session's preview open.
func (h *Hub) askPreviewDOM(ctx context.Context, sessionID, op, ref, value string) (string, error) {
	if h.previewDOM == nil {
		return "", fmt.Errorf("preview control is not available")
	}
	if h.preview == nil || h.preview.URL(sessionID) == "" {
		return "", fmt.Errorf("this session has no dev server running, so there is no page to inspect")
	}

	ask := protocol.PreviewDOMAsk{
		RequestID: "pdom_" + randToken(),
		SessionID: sessionID,
		Op:        op,
		Ref:       ref,
		Value:     value,
	}
	ch := h.previewDOM.open(ask.RequestID)
	defer h.previewDOM.close(ask.RequestID)

	// Broadcast rather than target a device: the daemon does not know which client, if any, has this
	// preview open. Every client sees the ask, only one can act on it, and the first answer wins.
	h.broadcast(protocol.TypePreviewDOMAsk, ask)

	select {
	case res := <-ch:
		if !res.OK {
			if res.Error != "" {
				return "", fmt.Errorf("%s", res.Error)
			}
			return "", fmt.Errorf("the preview could not complete that operation")
		}
		return res.Result, nil
	case <-time.After(previewDOMTimeout):
		// Overwhelmingly this means nobody has the page open, so say that rather than "timeout" —
		// the agent can act on the first and can do nothing with the second.
		return "", fmt.Errorf("no Iron Rain app is connected, so there is nothing that can render " +
			"this page. Open Iron Rain on any of your devices and try again")
	case <-ctx.Done():
		return "", ctx.Err()
	}
}

// previewBuiltinServer is the MCP server name these tools appear under.
const previewBuiltinServer = "ironrain-preview"

// previewTools describes the tools to the agent.
//
// The descriptions say what the tools CANNOT do as plainly as what they can. An agent that believes
// it can drive a page unattended will report a verification it never performed, which is worse than
// one that knows to ask for a person.
var previewTools = []mcp.BuiltinTool{
	{
		Name: "preview_snapshot",
		Description: "Look at the page your dev server is currently rendering, as a structured outline " +
			"of its interactive and text-bearing elements, each with a ref you can pass to preview_click " +
			"or preview_fill. This reads the LIVE DOM after the page's own JavaScript has run, so it " +
			"reflects what a person actually sees. Requires the Iron Rain app to be running somewhere — " +
			"it renders the page for me, without showing anything, if nobody has the preview open. If " +
			"no app is connected it will tell you so; do not guess at the page in that case.",
		InputSchema: json.RawMessage(`{"type":"object","properties":{},"additionalProperties":false}`),
	},
	{
		Name: "preview_click",
		Description: "Click an element in the live preview, identified by a ref from a previous " +
			"preview_snapshot. Take a fresh snapshot afterwards to see what changed.",
		InputSchema: json.RawMessage(`{"type":"object","properties":{` +
			`"ref":{"type":"string","description":"an element ref from preview_snapshot, e.g. e12"}},` +
			`"required":["ref"],"additionalProperties":false}`),
	},
	{
		Name: "preview_fill",
		Description: "Type a value into an input or textarea in the live preview, identified by a ref " +
			"from a previous preview_snapshot. Fires the input and change events a framework needs to " +
			"notice the value.",
		InputSchema: json.RawMessage(`{"type":"object","properties":{` +
			`"ref":{"type":"string","description":"an element ref from preview_snapshot, e.g. e12"},` +
			`"value":{"type":"string","description":"the text to type"}},` +
			`"required":["ref","value"],"additionalProperties":false}`),
	},
}

// RegisterPreviewTools installs the builtin server on the MCP manager, making the preview tools
// visible to every session's agent.
func (h *Hub) RegisterPreviewTools(mgr *mcp.Manager) {
	if mgr == nil {
		return
	}
	mgr.RegisterBuiltin(previewBuiltinServer, previewTools, h.callPreviewTool)
}

// callPreviewTool executes one of the preview tools for the calling session.
//
// The session is taken from the CONTEXT — from the bearer token the gateway resolved — never from an
// argument. An agent cannot name a session, so it cannot reach another one's page; that is not a
// check that could be got wrong, it is an input that does not exist.
func (h *Hub) callPreviewTool(ctx context.Context, tool string, args json.RawMessage) (json.RawMessage, error) {
	token := mcp.SessionTokenFrom(ctx)
	sessionID, ok := h.mcpTokens.session(token)
	if !ok {
		return nil, fmt.Errorf("these tools act on the calling session's own preview, and this call " +
			"has no session attached to it")
	}

	var p struct {
		Ref   string `json:"ref"`
		Value string `json:"value"`
	}
	if len(args) > 0 {
		_ = json.Unmarshal(args, &p)
	}

	var op string
	switch tool {
	case "preview_snapshot":
		op = "snapshot"
	case "preview_click":
		if p.Ref == "" {
			return nil, fmt.Errorf("preview_click needs a ref from a preview_snapshot")
		}
		op = "click"
	case "preview_fill":
		if p.Ref == "" {
			return nil, fmt.Errorf("preview_fill needs a ref from a preview_snapshot")
		}
		op = "fill"
	default:
		return nil, fmt.Errorf("unknown tool %q", tool)
	}

	out, err := h.askPreviewDOM(ctx, sessionID, op, p.Ref, p.Value)
	if err != nil {
		return nil, err
	}
	log.Printf("preview: %s for session %s", op, sessionID)
	return json.RawMessage(out), nil
}
